crates/router/src/workflows/revenue_recovery.rs

```
@@ -485,8 +546,334 @@ impl From<InternalDeciderRequest> for external_grpc_client::DeciderRequest {
+            total_retry_count_within_network: internal_request.total_retry_count_within_network,
             first_error_msg_time: internal_request.first_error_msg_time,
             wait_time: internal_request.wait_time,
         }
     }
 }
+
+#[cfg(feature = "v2")]
+#[derive(Debug, Clone)]
+pub struct ScheduledToken {
+    pub token_details: PaymentProcessorTokenDetails,
+    pub schedule_time: time::PrimitiveDateTime,
+}
+
+#[cfg(feature = "v2")]
+pub async fn get_best_psp_token_available(
+    state: &SessionState,
+    merchant_id: &id_type::MerchantId,
+    connector_customer_id: &str,
+    req_state: &ReqState,
+    profile: &domain::Profile,
+    payment_id: &id_type::GlobalPaymentId,
+    merchant_context: domain::MerchantContext,
+) -> Result<Option<ScheduledToken>, errors::ProcessTrackerError> {
+    //  Lock using payment_id
+    let locked =
+        RedisTokenManager::lock_connector_customer_status(state, connector_customer_id, payment_id)
+            .await?;
+
+    match locked {
+        true => Ok(None),
+
+        false => {
+            // Get existing tokens from Redis
+            let existing_tokens =
+                RedisTokenManager::get_connector_customer_payment_processor_tokens(
+                    state,
+                    connector_customer_id,
+                )
+                .await?;
+
+            // TODO: Insert into payment_intent_feature_metadata (DB operation)
+
+            let result = RedisTokenManager::get_tokens_with_retry_metadata(state, &existing_tokens);
+
+            let best_token_and_time = call_decider_for_payment_processor_tokens_select_closet_time(
+                state,
+                merchant_context,
+                merchant_id,
+                req_state,
+                profile,
+                &result,
+                payment_id,
+                connector_customer_id,
+            )
+            .await
+            .map_err(|_| errors::ProcessTrackerError::EApiErrorResponse)?;
```

why discarding error? change_context to EApiErrorResponse